scalanet/src/io/iohk/scalanet/peergroup/dynamictls/DynamicTLSPeerGroup.scala

```diff
@@ -0,0 +1,395 @@
+
+    val messageSubject = ConnectableSubject[M]()
+    val sslEngine = sslHandler.engine()
+
+    nettyChannel
+      .pipeline()
+      .addLast("ssl", sslHandler) //This needs to be first
+      .addLast(new ChannelInboundHandlerAdapter() {
+        override def userEventTriggered(ctx: ChannelHandlerContext, evt: Any): Unit = {
+          evt match {
+            case e: SslHandshakeCompletionEvent =>
+              val localAddress = InetMultiAddress(ctx.channel().localAddress().asInstanceOf[InetSocketAddress])
+              val remoteAddress = InetMultiAddress(ctx.channel().remoteAddress().asInstanceOf[InetSocketAddress])
+              if (e.isSuccess) {
+                // Hacky way of getting id of incoming peer form ssl engine. Only works becouse we have new engine for
+                // each connection
+                val params = sslEngine.getSSLParameters.getServerNames.asScala.head
+                val peerId = BitVector(params.getEncoded)
+                log.debug(
+                  s"Ssl Handshake server channel from $localAddress " +
+                    s"to $remoteAddress with channel id ${ctx.channel().id} and ssl status ${e.isSuccess}"
+                )
+                val channel = new ServerChannelImpl[M](nettyChannel, peerId, codec, messageSubject)
+                serverSubject.onNext(ChannelCreated(channel))
+              } else {
+                // Handshake failed we do not habe id of remote peer
+                serverSubject
+                  .onNext(
+                    PeerGroup.ServerEvent
+                      .HandshakeFailed(new HandshakeException(PeerInfo(BitVector.empty, remoteAddress), e.cause()))
+                  )
+              }
+          }
+        }
+      })
+      .addLast(new MessageNotifier(messageSubject, codec))
+  }
+
+  private[scalanet] class ServerChannelImpl[M](
+      val nettyChannel: SocketChannel,
+      peerId: BitVector,
+      codec: StreamCodec[M],
+      messageSubject: ConnectableSubject[M]
+  )(implicit scheduler: Scheduler)
+      extends Channel[PeerInfo, M] {
+
+    private val log = LoggerFactory.getLogger(getClass)
+
+    log.debug(
+      s"Creating server channel from ${nettyChannel.localAddress()} to ${nettyChannel.remoteAddress()} with channel id ${nettyChannel.id}"
+    )
+
+    override val to: PeerInfo = PeerInfo(peerId, InetMultiAddress(nettyChannel.remoteAddress()))
+
+    override def sendMessage(message: M): Task[Unit] = {
+      Task.fromTry(codec.encode(message).toTry).flatMap { enc =>
+        toTask(nettyChannel.writeAndFlush(Unpooled.wrappedBuffer(enc.toByteBuffer)))
+          .onErrorRecoverWith {
+            case e: IOException =>
+              Task.raiseError(new ChannelBrokenException[PeerInfo](to, e))
+          }
+      }
+    }
+
+    override def in: ConnectableObservable[M] = messageSubject
+
+    override def close(): Task[Unit] = {
+      messageSubject.onComplete()
+      toTask(nettyChannel.close())
+    }
+  }
+
+}
```

Comment:
can you split somehow this file? There is a lot of class implementation which are hard to read